lib/ring_logger/viewer.ex

```diff
@@ -0,0 +1,433 @@
+defmodule RingLogger.Viewer do
+  @moduledoc """
+  A Terminal UI (TUI) based viewer to browse a snapshot of the RingLogger buffer.
+  Offers a variety of filtering tools.
+
+  Start the viewer using the following line at the IEx prompt:
+
+  ```
+  RingLogger.viewer()
+  ```
+
+  Type `help`, `h` or `?` at the log viewer prompt to open a help page.
+  Type `exit` or `e` to exit the viewer.
+  """
+
+  use GenServer
+
+  @headers ["#", "Lvl", "App", "Msg", "Time"]
+  @header_lines 2
+  @footer_lines 2
+  @width_of_layout_items 55
+  @min_log_width 30
+  @min_log_entries 10
+
+  @min_height @header_lines + @footer_lines + @min_log_entries
+  @min_width @width_of_layout_items + @min_log_width
+
+  @init_state %{
+    current_screen: :list,
+    current_page: 0,
+    screen_dims: %{w: 0, h: 0},
+    lowest_log_level: nil,
+    before_boot: true,
+    grep_filter: nil,
+    applications_filter: [],
+    raw_logs: []
+  }
+
+  @level_strings ["emergency", "alert", "critical", "error", "warning", "notice", "info", "debug"]
+
+  @spec start_link(keyword()) :: :ignore | {:error, any()} | {:ok, pid()}
+  def start_link(options \\ []) do
+    GenServer.start_link(__MODULE__, options, name: __MODULE__)
+  end
+
+  @impl GenServer
+  def init(_options) do
+    screen_dims = get_screen_dims()
+
+    if screen_dims.w <= @min_width do
+      raise "Sorry, your terminal needs to be at least #{@min_width} columns wide to use this tool!"
+    end
+
+    if screen_dims.h <= @min_height do
+      raise "Sorry, your terminal needs to be at least #{@min_height} rows high to use this tool!"
+    end
+
+    Process.send_after(self(), :draw, 150)
+    IO.puts("Starting RingLogger Viewer...")
+
+    state = @init_state |> get_log_snapshot()
+
+    {:ok,
+     %{
+       state
+       | screen_dims: screen_dims
+     }}
+  end
+
+  #### Drawing and IO Functions
+
+  @impl GenServer
+  def handle_info(:draw, state) do
+    screen_dims = get_screen_dims()
+    new_state = %{state | screen_dims: screen_dims} |> do_draw()
+    Process.send_after(self(), :draw, 150)
+    {:noreply, new_state}
+  end
+
+  defp get_screen_dims() do
+    {:ok, rows} = :io.rows()
+    {:ok, cols} = :io.columns()
+
+    %{w: cols, h: rows}
+  end
+
+  defp do_draw(state) do
+    reset_screen(state.screen_dims)
+    print_header(state.screen_dims)
+
+    filtered_logs =
+      state.raw_logs
+      |> paginate_logs(state)
+
+    _ =
+      Enum.with_index(filtered_logs, fn entry, idx ->
+        print_log(state.screen_dims, entry, idx)
+      end)
+
+    print_footer(state.screen_dims)
+
+    case IO.gets(compute_prompt(state)) do
+      {:error, :interrupted} ->
+        Process.exit(self(), :normal)
+        state
+
+      string ->
+        string
+        |> to_string()
+        |> String.trim()
+        |> String.downcase()
```

Comment:
This downcase has a fun effect on grep which is case sensitive.

Reply:
I think that it would be ok to remove. The handlers below could be modified to be case insensitive on the first character.

Reply:
Yup, this downcase is in the wrong spot -- will move